pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -0,0 +1,256 @@
+}
+
+var _ capabilities.TriggerCapability = (*MercuryTriggerService)(nil)
+
+func NewMercuryTriggerService() *MercuryTriggerService {
+	return &MercuryTriggerService{
+		CapabilityInfo:        mercuryInfo,
+		chans:                 map[string]chan<- capabilities.CapabilityResponse{},
+		feedIDsForTriggerID:   make(map[string][]int64),
+		triggerIDsForFeedID:   make(map[int64][]string),
+		triggerIDToWorkflowID: make(map[string]string),
+	}
+}
+
+// maybe use mercury-pipline reports.go report struct instead of MercuryReport struct, and then we can use the pipeline report generators
+func (o *MercuryTriggerService) ProcessReport(reports []mercury.FeedReport) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	currentTime := time.Now()
+	unixTimestampMillis := currentTime.UnixNano() / int64(time.Millisecond)
+	triggerIDsForReports := make(map[string][]int)
+	reportIndex := 0
+
+	for _, report := range reports {
+		for _, triggerID := range o.triggerIDsForFeedID[report.FeedID] {
+			// if its not initialized, initialize it
+			if _, ok := triggerIDsForReports[triggerID]; !ok {
+				triggerIDsForReports[triggerID] = make([]int, 0)
+			}
+			triggerIDsForReports[triggerID] = append(triggerIDsForReports[triggerID], reportIndex)
+		}
+		reportIndex++
+	}
+
+	// Then for each trigger id, find which reports correspond to that trigger and create an event bundling the reports
+	// and send it to the channel associated with the trigger id.
+	for triggerID, reportIDs := range triggerIDsForReports {
+		reportPayload := make([]mercury.FeedReport, 0)
+		for _, reportID := range reportIDs {
+			reportPayload = append(reportPayload, reports[reportID])
+		}
+
+		triggerEvent := mercury.TriggerEvent{
+			TriggerType: "mercury",
+			ID:          GenerateTriggerEventID(reportPayload),
+			Timestamp:   strconv.FormatInt(unixTimestampMillis, 10),
+			Payload:     reportPayload,
+		}
+
+		val, err := mercury.Codec{}.WrapMercuryTriggerEvent(triggerEvent)
+		if err != nil {
+			return err
+		}
+
+		// Create a new CapabilityResponse with the MercuryTriggerEvent
+		capabilityResponse := capabilities.CapabilityResponse{
+			Value: val,
+		}
+
+		wfID := o.triggerIDToWorkflowID[triggerID]
+		ch, ok := o.chans[wfID+triggerID]
+		if ok {
+			ch <- capabilityResponse
+		} else {
+			return fmt.Errorf("no registration for %s", triggerID)
+		}
+	}
+	return nil
+}
+
+func (o *MercuryTriggerService) RegisterTrigger(ctx context.Context, callback chan<- capabilities.CapabilityResponse, req capabilities.CapabilityRequest) error {
+	wid := req.Metadata.WorkflowID
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	triggerID, err := o.GetTriggerID(req)
+	if err != nil {
+		return err
+	}
+	// If triggerId is already registered, return an error
+	if _, ok := o.chans[wid+triggerID]; ok {
+		return fmt.Errorf("triggerId %s already registered", triggerID)
+	}
+	feedIDs := o.GetFeedIDs(req) // TODO: what if feedIds is empty? should we throw an error or allow it?
+
+	o.chans[wid+triggerID] = callback
```

Comment:
Would it simplify things if we just identified a trigger by (wid, triggerID) throughout the whole file?
Or maybe just assume and validate that triggerIDs are globally unique like @HenryNguyen5 suggested earlier (IIRC)?

Reply:
If we assumed that triggers were globally unique, then you do lose a little security in that right now, workflow A can't remove workflow Bs triggers since they are effectively namespaced. I'm okay with identifying the trigger as (wid, triggerID) throughout the file though if that simplifies the code throughout (which I think it would).